REGRESSION(299318@main): [GStreamer][MSE] Audio playback broken on instagram

philn opened the pull request:
#### 4a3ebdf6bf5780c83d77cb52a7843d994da7a3eb
<pre>
REGRESSION(299318@main): [GStreamer][MSE] Audio playback broken on instagram
<a href="https://bugs.webkit.org/show_bug.cgi?id=299252">https://bugs.webkit.org/show_bug.cgi?id=299252</a>

Reviewed by Alicia Boya Garcia.

299318@main introduced a performance regression and also broke Instagram stories audio playback.
Instead of forcing ADTS framing we now rewrite the codec_data supplied by the demuxer only if the
underlying AAC audio_object_type was set to SBR or PS, which the avdec_aac decoder doesn&apos;t handle
very well. When any of those object types is detected, we replace it with an LC (Low Complexity)
identifier and re-use the existing sampling frequency and channels bits from the previous
codec_data. avdec_aac handles LC well and there is no noticeable rendering issue on Kaltura and
Instagram stories anymore.

Thanks to Alicia Boya Garcia &lt;aboya@igalia.com&gt; for her precious help during the investigation.

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::aacGetAudioObjectType):
(WebCore::createOptionalParserForFormat):

Canonical link: <a href="https://commits.webkit.org/301209@main">https://commits.webkit.org/301209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fb454f49b3b4d81a741500191e76f9a1e19fae1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125248 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44915 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35654 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132099 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77117 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127125 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45606 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53476 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95362 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128202 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36413 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112002 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75901 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35311 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30170 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75578 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106180 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30395 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134786 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52057 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/39836 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103829 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52492 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108223 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103597 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26382 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48946 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27235 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49153 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51949 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57728 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51313 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54667 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53005 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -883,16 +900,71 @@ GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin,
+            }
+
+            GST_DEBUG_OBJECT(bin, "AAC SBR or PS object type detected in codec data, replacing it with a LC object type. The decoder should be able to handle this properly");
+            auto capsSetter = makeGStreamerElement("capssetter"_s);
```
Important information from the codec-data (such as the inner sample rate or whether SBR is used) may change during a quality change, so we shouldn't set a fixed codec-data permanently.

Instead we can add a probe do the codec-data swapping when we get a caps event.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -883,16 +900,71 @@ GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin,
+            if (!capsSetter)
+                break;
+
+            int channels = gstStructureGet<int>(structure, "channels"_s).value_or(1);
```
Since we're trying to do implicit signalling we need to get the amount of channels of the inner AAC LC stream. That is given to us in the 4 bits following the sampling frequency index.

It is very important that we get those bits from the codec-data and not from the caps structure.

The caps `"channels"` field in the caps is higher level than we need: it will report stereo for PS (HE AACv2) files that use explicit signalling. To do implicit signalling on the other hand, we need to use the number of channels from the inner AAC LC stream however, which in the case of HE AACv2 is mono.

Using the 4 bits from the codec-data will avoid this problem: whatever we read here we use in the the final codec-data.

Here are [some test vectors](https://www2.iis.fraunhofer.de/AAC/stereo.html) you can use to confirm this:
* https://www2.iis.fraunhofer.de/AAC/SBRtestStereoAot29Sig2.mp4 PS explicit hierarchical signalling
* https://www2.iis.fraunhofer.de/AAC/SBRtestStereoAot29Sig0.mp4 PS implicit signalling

philn replied:
OK, I'm using `gst_codec_utils_aac_get_channels()` to get this info now.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -883,16 +900,71 @@ GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin,
-            if (result)
+            // Make sure the decoder will receive an LC bitstream as input, otherwise bitstreams
+            // advertized as SBR or PS might not be handled correctly by some decoders such as
+            // avdec_aac.
```
More precise wording:

Replace explicit signalling of SBR (HE AAC) and PS (HE AACv2) with implicit signalling (i.e. let the decoder figure out what AAC extensions are used).
Unfortunately it is not uncommon for AAC streams to explicitly signal extensions they don't actually use, which is known to cause problems in avdec_aac.

ntrrgc reviewed (changes requested):
Aside from what I commented, LGTM.

There is one case of explicit signalling we're not handling here: Backwards-compatible signalling. In that case, AOT=2 appears first, and the SBR and PS extensions appear later in the codec-data. It's a bit trickier to parse because it requires parsing GASpecificInfo as well. For now we can let it be, and if we actually encounter a file where that is a problem, handle it then.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -883,16 +899,73 @@ GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin,
+            if (!result)
+                break;
+
+            gst_pad_add_probe(demuxerSrcPad.get(), GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM, reinterpret_cast<GstPadProbeCallback>(+[]([[maybe_unused]] GstPad* pad, GstPadProbeInfo* info, gpointer) -> GstPadProbeReturn {
```
nit: this is too big of a lambda and would be better off as a separate static function. You could call it `aacSbrForceImplicitSignalling` or something of the like.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -883,16 +899,73 @@ GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin,
+
+                GST_DEBUG_OBJECT(pad, "AAC SBR or PS object type detected in codec data, replacing it with a LC object type. The decoder should be able to handle this properly");
+
+                auto channels = gst_codec_utils_aac_get_channels(data.data(), data.size());
```
I was concerned that gst_codec_utils_aac_get_channels() could be too high level and actually parse PS as well, but it doesn't. So as long as it is not changed afterwards, this would work.

Personally I would have just read the bits myself since it's not more difficult than reading the frequency bits.

philn replied:
Well I think we can consider that pbutils API behaviour stable, its doc clearly implies the value return should conform with the spec.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -883,16 +899,73 @@ GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin,
+
+                GRefPtr<GstBuffer> codecDataBuffer;
+                gst_structure_get(structure, "codec_data", GST_TYPE_BUFFER, &codecDataBuffer.outPtr(), nullptr);
+                GstMappedBuffer data(codecDataBuffer, GST_MAP_READ);
```
Also return early with GST_PAD_PROBE_OK and a warning if the length of codec_data < 2, since otherwise the only length checks are asserts, which are not compiled in release.

ntrrgc reviewed (approved):
LGTM with nits.

ntrrgc reviewed (approved):
r+

philn committed:
REGRESSION(299318@main): [GStreamer][MSE] Audio playback broken on instagram
https://bugs.webkit.org/show_bug.cgi?id=299252

Reviewed by Alicia Boya Garcia.

299318@main introduced a performance regression and also broke Instagram stories audio playback.
Instead of forcing ADTS framing we now rewrite the codec_data supplied by the demuxer only if the
underlying AAC audio_object_type was set to SBR or PS, which the avdec_aac decoder doesn't handle
very well. When any of those object types is detected, we replace it with an LC (Low Complexity)
identifier and re-use the existing sampling frequency and channels bits from the previous
codec_data. avdec_aac handles LC well and there is no noticeable rendering issue on Kaltura and
Instagram stories anymore.

Thanks to Alicia Boya Garcia <aboya@igalia.com> for her precious help during the investigation.

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::aacGetAudioObjectType):
(WebCore::createOptionalParserForFormat):

Canonical link: https://commits.webkit.org/301209@main
